diff --git a/config.h b/config.h
index af199565..873627e0 100644
--- a/config.h
+++ b/config.h
@@ -60,7 +60,9 @@
 #define HAVE_DECL_STRERROR_R 1
 
 /* Define to 1 if you have the <error.h> header file. */
+#if defined(__BIONIC__) || defined(__GLIBC__)
 #define HAVE_ERROR_H 1
+#endif
 
 /* Define to 1 if you have the <err.h> header file. */
 #define HAVE_ERR_H 1
@@ -69,7 +71,7 @@
 #define HAVE_FALLTHROUGH 1
 
 /* Defined if __attribute__((gcc_struct)) is supported */
-#define HAVE_GCC_STRUCT 1
+/* #undef HAVE_GCC_STRUCT */
 
 /* Define to 1 if you have the `getrlimit' function. */
 #define HAVE_GETRLIMIT 1
@@ -167,7 +169,9 @@
 #define STDC_HEADERS 1
 
 /* Define to 1 if strerror_r returns char *. */
+#if defined(__GLIBC__) || defined(__BIONIC__)
 #define STRERROR_R_CHAR_P 1
+#endif
 
 /* Support bzip2 decompression via -lbz2. */
 /* #undef USE_BZLIB */
